dipy/tracking/local/localtrack.pyx

```diff
@@ -208,3 +199,296 @@ cdef int _local_tracker(
+        Temporary array for the weights of particles.
+    particle_states : array, float, (2, particle_count, 2)
+        Temporary array for the states of particles.
+
+    Returns
+    -------
+    end : int
+        Length of the tracked streamline
+    tissue_class : TissueClass
+        Ending state of the streamlines as determined by the TissueClassifier.
+
+    """
+    cdef:
+        size_t i
+        TissueClass tissue_class
+        double dir[3]
+        double vs[3]
+        double seed[3]
+
+    if (seed_pos.shape[0] != 3 or first_step.shape[0] != 3 or
+            voxel_size.shape[0] != 3 or streamline.shape[1] != 3):
+        raise ValueError('Invalid input parameter dimensions.')
+
+    for i in range(3):
+        dir[i] = first_step[i]
+        vs[i] = voxel_size[i]
+        seed[i] = seed_pos[i]
+
+    i = _pft_tracker(dg, tc, seed, dir, vs, streamline,
+                     directions, step_size, &tissue_class, pft_max_nbr_back_steps,
+                     pft_max_steps, pft_max_trials, particle_count,
+                     particle_paths, particle_dirs, particle_weights,
+                     particle_states)
+    return i, tissue_class
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+@cython.cdivision(True)
+cdef _pft_tracker(DirectionGetter dg,
+                  ConstrainedTissueClassifier tc,
+                  double* seed,
+                  double* dir,
+                  double* voxel_size,
+                  np.float_t[:, :] streamline,
+                  np.float_t[:, :] directions,
+                  double step_size,
+                  TissueClass * tissue_class,
+                  int pft_max_nbr_back_steps,
+                  int pft_max_steps,
+                  int pft_max_trials,
+                  int particle_count,
+                  np.float_t[:, :, :, :] particle_paths,
+                  np.float_t[:, :, :, :] particle_dirs,
+                  np.float_t[:] particle_weights,
+                  np.int_t[:, :, :] particle_states):
+    cdef:
+        int i, pft_trial, pft_streamline_i, pft_nbr_back_steps, strl_array_len
+        int pft_nbr_steps
+        double point[3]
+        double voxdir[3]
+        void (*step)(double* , double*, double) nogil
+
+    copypoint(seed, point)
+    copypoint(seed, &streamline[0,0])
+    copypoint(dir, &directions[0, 0])
+
+    tissue_class[0] = TRACKPOINT
+    pft_trial = 0
+    i = 1
+    strl_array_len = streamline.shape[0]
+    while i < strl_array_len:
+        if dg.get_direction_c(point, dir):
+            # no valid diffusion direction to follow
+            tissue_class[0] = INVALIDPOINT
+        else:
+            for j in range(3):
+                voxdir[j] = dir[j] / voxel_size[j]
+
+            fixed_step(point, voxdir, step_size)
+            copypoint(point, &streamline[i, 0])
+            copypoint(dir, &directions[i, 0])
+            tissue_class[0] = tc.check_point_c(point)
+            i += 1
+
+        if tissue_class[0] == TRACKPOINT:
```

Comment:
I tested more and I don't understand what's going on. I  added a print on line 347

    print 'Loop', i, point, class

and I obtain

    Loop 2 0.669794495731 2.11007159921 1.86826715409 OUTSIDEIMAGE
    Loop 3 0.589588991463 1.97014319842 1.98653430818 OUTSIDEIMAGE
    Loop 4 0.57799543161 1.82547363239 2.12414355945 OUTSIDEIMAGE
    Loop 5 0.566401871757 1.68080406636 2.26175281072 OUTSIDEIMAGE
    Loop 6 0.554808311905 1.53613450033 2.39936206199 OUTSIDEIMAGE
    Loop 7 0.543214752052 1.3914649343 2.53697131326 ENDPOINT
    END ENDPOINT 7
    Loop 2 0.830205504269 2.38992840079 1.63173284591 OUTSIDEIMAGE
    Loop 3 0.910411008537 2.52985680158 1.51346569182 INVALIDPOINT
    Starting pft on 3 8
    ...

I checked and all `OUTSIDEIMAGE` are in the final streamline. Why are they considered valid if they are classified as Outside the image?

Reply:
`OUTSIDEIMAGE` is equal to `ENDPOINT` for PFT. Since we don't have information at that point, we stop the tracking and keep the streamline. This is especially important to keep streamlines exiting the image volume through the spinal cord. 

There is bug here as the tracking should stop when it reach `OUTSIDEIMAGE`. Thank you for pointing this out. I fixed it and I added comments to clarify why the algorithm reaches a `break` or a `continue` staetment.

Reply:
@gabknight I think we had a similar discussion in the prototype version of this code: you say you want to keep streamlines exiting the image volume through the spinal cord. So, is the chosen behavior for now to keep the additional point? If so, won't that bring back some issues when trying to project the streamlines to a grid / volume? I.e. the point outside won't match a possible voxel coordinate, and might cause undefined behavior?

Reply:
@jchoude I don't keep the `OUTSIDEIMAGE` point. My answer was misleading. `ENDPOINT` and `OUTSIDEIMAGE` show  **almost** equal behaviors, except at line 381 (localtrack.pyx), where the last point is removed for the `OUTSIDEIMAGE` case. I'm not sure we test that all points are within the image volume. I will add this test.

Reply:
Oh ok all good then, I just misunderstood.

Thanks a lot for all this work!